design/KEPS/graph-cache/proposal.md

```diff
@@ -0,0 +1,236 @@
+  - Benefits:
+    - favorite graphs always ready
+  - Downsides:
+    - it may be very unlikely for users to want to define graph queries in config
+    - config-based queries are fixed may become dated
+    - updates require kiali pod restarts
+    - some users may not have permissions to view the data resulting from the predefined graph
+    - **CRITICAL**: Configuration barrier makes this impractical for real-world usage
+    - **CRITICAL**: Permission mismatch between controller SA and users has no good solution
+
+- **Background refresh (re-compute)** - ✅ **RECOMMENDED (Nov 2025)**
+
+  - Like today, users can set graph options as needed, and then launch a graph request. The user will wait for the initial graph to render. The result will be cached and the server will proactively re-compute the same graph, updating the cache. Requests for the same graph will return the most recent cached result. A change of graph, a cache-hit timeout, or a session end would end the re-compute.
+  - Benefits:
+    - avoids permission issues and config, works with any graph at any time
+    - no user configuration required (automatic, self-service)
+    - permission-aware caching (per-session)
+    - self-optimizing (popular graphs stay cached)
+  - Downsides:
+    - still have to wait on the first request and any subsequent change to the requested graph
+
+# Proposed Implementation
+
+## Architecture
+
+The graph cache implementation uses a **per-session** caching strategy with background refresh jobs. Unlike the controller-based approach used for Validations and Health, the graph cache uses a custom in-memory cache for several key reasons:
+
+### Why Not a Kubernetes Controller?
+
+1. **Non-Kubernetes Data Sources**: Graph generation relies primarily on Prometheus metrics, not Kubernetes objects. There are no K8s resources to watch that would trigger graph regeneration.
+
+2. **Time-Based Refresh**: Graphs must refresh based on time window advancement (e.g., "last 5 minutes"), not configuration changes. This requires periodic polling rather than event-driven updates.
+
+3. **Permission Context**: Graph data must be generated using the user's permissions (RBAC), not the Kiali service account. A controller running with elevated privileges would bypass user access controls.
+
+4. **Dynamic Configuration**: Graph options (namespaces, graph type, display options) are user-specified at request time, not pre-defined in configuration. This makes them unsuitable for controller-based pre-computation.
+
+5. **Session Lifecycle**: Graphs are ephemeral, tied to user sessions. They should be evicted when users navigate away, not persist indefinitely like other cached data.
+
+## Per-Session Caching
+
+Each user session maintains its own cached graph, uniquely identified by `sessionID`. This approach:
+
+- **Preserves permissions**: Each session uses the requesting user's auth token for Prometheus queries
```

Comment:
I think this is different from how we do prom queries today. We use the single creds configured in `external_services.prometheus`. Is this adding per user prom creds?

Reply:
@jshaughn can you clarify this point? iirc we use global prom creds every time we do a query. We don't use per user creds. Although that might be a nice enhancement in the future.